package/MDAnalysis/core/topology.py

```diff
@@ -17,13 +17,58 @@
 Topology object --- :mod:`MDAnalysis.core.topology'
 ===================================================================
 
+
+"""
+
+"""
+# second docstring == multiline comment
+
+TODO Notes:
+  Could make downshift tables lazily built! This would
+    a) Make these not get built when not used
+    b) Optimise moving multiple atoms between residues as only built once afterwards
+
+  Could optimise moves by only updating the two parent tables rather than rebuilding everything!
+
+
 """
+from six.moves import zip
 import numpy as np
 
 from ..lib.mdamath import one_to_many_pointers
 from .topologyattrs import Atomindices, Resindices, Segindices
 
 
+def make_downshift_arrays(upshift):
+    """From an upwards translation table, create the opposite direction
+
+    Turns a many to one mapping to a one to many mapping
+
+    Eg :: 
+      AR = np.array([0, 1, 0, 2, 2, 0, 2])
+    
+      make_downshift_arrays(AR)
+      array([array([0, 2, 5]), array([1]), array([3, 4, 6])], dtype=object)
+
+    So entry 0 informs that items 0, 2 & 5 all belong to group 0.
+
+    Returns
+    -------
+    Numpy array of numpy arrays (dtype object)
+    Length : 
+    """
+    order = np.argsort(upshift)
+
+    upshift_sorted = upshift[order]
+    borders = [None] + list(np.nonzero(np.diff(upshift_sorted))[0] + 1) + [None]
+
+    # returns an array of arrays
+    # Add None to end of array to force it to be of type Object
+    # Without this, a rectangular array gets squashed into a single array
+    return np.array([np.array(order[x:y], copy=True, dtype=np.int)
```

Comment:
So if the arrays are all the same length (and can be make 2d)
```python
In [2]: a1 = np.arange(3)

In [3]: a2 = np.arange(3) + 1

In [4]: a3 = np.arange(3) + 2

In [5]: b = np.array([a1, a2, a3], dtype=object)

In [6]: b
Out[6]: 
array([[0, 1, 2],
       [1, 2, 3],
       [2, 3, 4]], dtype=object)

In [11]: b[0].dtype
Out[11]: dtype('O')
```
So each member array gets made into dtype object

But if they can't be made 2d
```python
In [12]: a2 = np.arange(4) + 1

In [13]: b = np.array([a1, a2, a3], dtype=object)

In [14]: b[0].dtype
Out[14]: dtype('int64')
```
Then each member array retains its original dtype (the desired behaviour!)

I think what's happening is the potential squashing into a single array happens before any dtype arguments are taken into account.